[SymmMem] find_path does not search /usr/local/lib

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157695
* #157513

This PR uses `find_library` to replace `find_path`. 
It also searches for NVSHMEM host lib and device lib separately. 

Tested against system install location: /usr/local/lib and /usr/local/include.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

kwen2501 committed:
[SymmMem] find_path does not search /usr/local/lib

[ghstack-poisoned]

kwen2501 committed:
Update on "[SymmMem] find_path does not search /usr/local/lib"




[ghstack-poisoned]

Skylion007 commented:
@kwen2501 Okay good, we have some compilation breakage on master to sort it. Missing includes perhaps?

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 commented on caffe2/CMakeLists.txt:
```diff
@@ -999,20 +999,28 @@ elseif(USE_CUDA)
+      PATHS $ENV{NVSHMEM_HOME}/lib ${NVSHMEM_PY_DIR}/lib
+      DOC "The location of NVSHMEM device library.")
     find_path(NVSHMEM_INCLUDE_DIR
       NAMES nvshmem.h
```
Yeah, it seems NVSHMEM would include `nvshmemx_collective_launch` from nvshmem.h for some higher SM arch's, but not for some lower ones (the failed test runs on SM 52). Now I added `#include <nvshmemx.h>` for the lower SMs.

kwen2501 committed:
Update

[ghstack-poisoned]

Skylion007 commented on cmake/Summary.cmake:
```diff
@@ -172,7 +172,7 @@ function(caffe2_print_configuration_summary)
     message(STATUS "    USE_SYSTEM_NCCL     : ${USE_SYSTEM_NCCL}")
   endif()
-  message(STATUS "  NVSHMEM_LIB_DIR       : ${NVSHMEM_LIB_DIR}")
+  message(STATUS "  Found NVSHMEM         : ${NVSHMEM_INCLUDE_DIR}")
```
This is printed even if   `message(STATUS "  Found NVSHMEM         : ${NVSHMEM_INCLUDE_DIR}")` is empty...

kwen2501 replied:
Empty means not found. 
It works on my machine tho:
```
...
--   USE_NCCL              : ON
--     USE_SYSTEM_NCCL     : OFF
--   Found NVSHMEM         : /usr/local/include
--   USE_NNPACK            : 0
--   USE_NUMPY             : ON
...
```

Skylion007 commented on caffe2/CMakeLists.txt:
```diff
@@ -999,20 +999,28 @@ elseif(USE_CUDA)
       # In pip install case, the lib suffix is `.so.3` instead of `.so`
-      NAMES libnvshmem_host.so libnvshmem_host.so.3
+      NAMES nvshmem_host nvshmem_host.so.3
       PATHS $ENV{NVSHMEM_HOME}/lib ${NVSHMEM_PY_DIR}/lib
```
Okay, I don't it actually found NVSHMEM with the ints for sumr reason...

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 commented on caffe2/CMakeLists.txt:
```diff
@@ -999,20 +999,30 @@ elseif(USE_CUDA)
+      DOC "The location of NVSHMEM host library.")
+    find_library(NVSHMEM_DEVICE_LIB
+      # Device lib is a `.a` file
+      NAMES nvshmem_device
+      HINTS $ENV{NVSHMEM_HOME} ${NVSHMEM_PY_DIR}
+      PATH_SUFFIXES lib lib64 cuda/lib cuda/lib64 lib/x64
+      DOC "The location of NVSHMEM device library.")
     find_path(NVSHMEM_INCLUDE_DIR
       NAMES nvshmem.h
-      PATHS $ENV{NVSHMEM_HOME}/include ${NVSHMEM_PY_DIR}/include
+      HINTS $ENV{NVSHMEM_HOME}/include ${NVSHMEM_PY_DIR}/include
```
We missed one replacement of PATHS -> HINTS. 
HINTS: searched before default paths.
PATHS: searched after default paths. 
Now it should be good. (I tested it locally)

kwen2501 commented:
@pytorchbot merge -f "Tests passed previously; changing from PATHS to HINTS to prioritize search"

kwen2501 commented:
@pytorchbot cherry-pick --onto release/2.8 -c critical

kwen2501 commented:
@pytorchbot revert

kwen2501 commented:
@pytorchbot revert -m "Changing it to be landable on 2.8 branch" -c landrace

kwen2501 committed:
Update on "[SymmMem] find_path does not search /usr/local/lib"


This PR uses `find_library` to replace `find_path`. 
It also searches for NVSHMEM host lib and device lib separately. 

Tested against system install location: /usr/local/lib and /usr/local/include.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot cherry-pick --onto cherry-pick-157513-by-pytorch_bot_bot_ -c critical
